### PR TITLE
Learning to prompt

### PR DESCRIPTION
**Pull request by prabhuteja12**
This PR introduces the learning to prompt method for continual learning for ViTs. It introduces two learners and corresponding updaters . 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/renate/benchmark/models/base.py [hunk 1]
```diff
@@ -87,7 +87,7 @@ def get_extra_state(self, encode=True) -> Any:
         Encode converts the state into a torch tensor so that Deepspeed serialization works.
         We don't encode any of the super() calls, but encode only the final dict.
         """
-        extra_state = super().get_extra_state(encode=not encode)
+        extra_state = super().get_extra_state(encode=False)
```
Comment (wistuba) [inline src/renate/benchmark/models/base.py:90]: why?

File: src/renate/benchmark/models/l2p.py [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+
+    def forward(self, x: torch.Tensor, manual_prompt_indices: Optional[torch.LongTensor] = None):
+        """
+        Args:
+            x: Image features extracted. It can be [CLS] token or something else.
+            manual_prompt_indices: Indices to manually select prompts from pool, instead of
+                selecting from
+        """
+        # Here x refers to the features already extracted. These can be [CLS] token representation
+        # or something else altogether. But it has to be of dimension B x self.pd."
```
Comment (wistuba) [inline src/renate/benchmark/models/l2p.py:84]: great candidate for docstring

File: src/renate/benchmark/models/vision_transformer.py [hunk 3]
```diff
@@ -8,37 +8,36 @@
...
             output_attentions,
             output_hidden_states,
             interpolate_pos_encoding,
             return_dict,
         )
-
         if isinstance(out_to_filter, BaseModelOutputWithPooling):
-            return out_to_filter.pooler_output
-        return out_to_filter[1]
+            return out_to_filter.pooler_output if cls_feat else out_to_filter.last_hidden_state
```
Comment (wistuba) [inline src/renate/benchmark/models/vision_transformer.py:39]: this doesn't work together with the evaluator. no idea why

File: src/renate/benchmark/models/vision_transformer.py [hunk 4]
```diff
@@ -84,8 +83,12 @@ def __init__(
     ) -> None:
         if pretrained_model_name_or_path:
             model = FeatureExtractorViTModel.from_pretrained(
-                pretrained_model_name_or_path=pretrained_model_name_or_path, return_dict=False
+                pretrained_model_name_or_path=pretrained_model_name_or_path,
+                return_dict=False,
+                add_pooling_layer=False,
             )
+
+            constructor_args = dict(pretrained_model_name_or_path=pretrained_model_name_or_path)
```
Comment (wistuba) [inline src/renate/benchmark/models/vision_transformer.py:91]: why is add_pooling_layer=False now? what does this change?

**Review by wistuba — CHANGES_REQUESTED** [review]
new updater needs to be added to docs

we need to check whether it works with benchmarking

merge #364 first

File: src/renate/benchmark/models/base.py (hunk 1 above)
Comment (prabhuteja12) [reply src/renate/benchmark/models/base.py:90]: Crashes when there are nested RenateBenchmarkingModules. The simplification is that all sub-modules are not encoded, and there is only one final encoding.

File: src/renate/benchmark/models/l2p.py (hunk 2 above)
Comment (prabhuteja12) [reply src/renate/benchmark/models/l2p.py:84]: already there.

File: src/renate/benchmark/models/l2p.py [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+
+        for p in self._backbone["vit"].parameters():
+            p.requires_grad = False
+        self._backbone["vit"].eval()
+        for p in self._backbone["prompter"].parameters():
+            p.requires_grad = True
+
+    def forward(self, x: torch.Tensor, task_id: str = defaults.TASK_ID) -> torch.Tensor:
+        with torch.no_grad():
+            prompt_pool_input = self._backbone["vit"].get_logits(x, cls_feat=False)
```
[the remark this answers is not in the document]
Comment (prabhuteja12) [reply src/renate/benchmark/models/l2p.py]: Where?

File: src/renate/benchmark/models/l2p.py [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+        train_prompt_keys: Whether to train the prompt keys. Currently unused.
+        similarity_fn: Similarity function between input features and prompt keys.
+        per_batch_prompt: Flag to use the same prompts for all elements in the batch.
+        prompt_embedding_features: Image feature type used to compute the similarity to prompt keys.
+        patch_pooler: Features to feed the classifier.
+    """
+
+    def __init__(
+        self,
+        pretrained_model_name_or_path="google/vit-base-patch16-224-in21k",
```
Comment (wistuba) [inline src/renate/benchmark/models/l2p.py:147]: any objection to make `google/vit-base-patch16-224` the default?

Comment (prabhuteja12) [reply src/renate/benchmark/models/l2p.py:147]: No

File: src/renate/benchmark/experiment_config.py [hunk 7]
```diff
@@ -65,6 +66,7 @@
     "VisionTransformerL32": VisionTransformerL32,
     "VisionTransformerH14": VisionTransformerH14,
     "HuggingFaceTransformer": HuggingFaceSequenceClassificationTransformer,
+    "PromptedVisionTransformer": PromptedVisionTransformer,
```
[the remark this answers is not in the document]
Comment (prabhuteja12) [reply src/renate/benchmark/experiment_config.py]: Now, we just specify a model name and it picks ViT or a text transformer.

File: src/renate/benchmark/experiment_config.py [hunk 8]
```diff
@@ -356,13 +365,23 @@ def train_transform(dataset_name: str, model_name: Optional[str] = None) -> Opti
...
         return None
     if dataset_name in ["CIFAR10", "CIFAR100"]:
-        return transforms.Compose(
-            [
-                transforms.RandomCrop(32, padding=4),
-                transforms.RandomHorizontalFlip(),
-                _get_normalize_transform(dataset_name),
-            ]
-        )
+        if (model_name is not None) and ("Prompted" in model_name):
```
Comment (wistuba) [inline src/renate/benchmark/experiment_config.py:368]: Can you include all ViT here?

File: src/renate/benchmark/experiment_config.py [hunk 9]
```diff
@@ -405,7 +424,15 @@ def test_transform(
     ] + wild_time_data.list_datasets() or dataset_name.startswith("hfd-"):
         return None
     if dataset_name in ["CIFAR10", "CIFAR100"]:
-        return _get_normalize_transform(dataset_name)
+        if (model_name is not None) and ("Prompted" in model_name):
```
Comment (wistuba) [inline src/renate/benchmark/experiment_config.py:427]: same as above

File: src/renate/benchmark/models/l2p.py [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+        prediction_strategy: Continual learning strategies may alter the prediction at train or test
+            time.
+        add_icarl_class_means: If ``True``, additional parameters used only by the
+            ``ICaRLModelUpdater`` are added. Only required when using that updater.
+        pool_size: Total size of the prompt pool.
+        pool_selection_size: Number of prompts to select from the pool.
+        prompt_size: Number of tokens each prompt is equivalent to.
+        prompt_key_dim: Dimensions of the prompt key used to compute similarity. It has to be same
+            to the dimensions of `x` in forward.
+        embedding_dim: Output dimension of the token/patch embedding layer
```
Comment (wistuba) [inline src/renate/benchmark/models/l2p.py:138]: `embedding_dim` does not exist as an argument. do you mean something else?

File: test/renate/updaters/test_learner.py [hunk 11]
```diff
@@ -39,7 +39,11 @@ def check_learner_variables(learner: Learner, expected_variable_values: Dict[str
...
-    checkpoint_dict = {}
-    learner.on_save_checkpoint(checkpoint=checkpoint_dict)
-    check_learner_variables(learner, checkpoint_dict)
+    if learner_class not in L2P_LEARNERS:
+        model, learner, learner_kwargs = get_model_and_learner_and_learner_kwargs(learner_class)
+        checkpoint_dict = {}
+        learner.on_save_checkpoint(checkpoint=checkpoint_dict)
+        check_learner_variables(learner, checkpoint_dict)
+    else:
+        with pytest.raises(AssertionError):
```
Comment (wistuba) [inline test/renate/updaters/test_learner.py:48]: can we have an equivalent test here?

**Review by wistuba — CHANGES_REQUESTED** [review]

File: src/renate/benchmark/models/l2p.py (hunk 10 above)
Comment (prabhuteja12) [reply src/renate/benchmark/models/l2p.py:138]: By error. This is inferred from the transformer's embedding dim.

File: test/renate/updaters/test_learner.py (hunk 11 above)
Comment (prabhuteja12) [reply test/renate/updaters/test_learner.py:48]: fixed in d1a67c1

File: src/renate/benchmark/models/l2p.py (hunk 10 above)
Comment (prabhuteja12) [reply src/renate/benchmark/models/l2p.py:138]: Fixed in 07605d18b8ecd3e1b6b28c8a69fc75c7e698caca

File: src/renate/benchmark/experiment_config.py (hunk 9 above)
Comment (prabhuteja12) [reply src/renate/benchmark/experiment_config.py:427]: Fixed c667119

File: src/renate/updaters/experimental/l2p.py [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+        self.prompt_sim_loss_weight = prompt_sim_loss_weight
+        self._loss_collections["train_losses"].update({"key_sim_loss": torchmetrics.MeanMetric()})
+
+    def training_step(
+        self, batch: Tuple[NestedTensors, torch.Tensor], batch_idx: int
+    ) -> STEP_OUTPUT:
+        loss_dict = super().training_step(batch, batch_idx=batch_idx)
+        key_similarity = -1 * self.prompt_sim_loss_weight * self._model.similarity_score
+        loss_dict["loss"] += key_similarity
+        self._loss_collections["train_losses"]["key_sim_loss"](-key_similarity)
```
[the remark this answers is not in the document]
Comment (prabhuteja12) [reply src/renate/updaters/experimental/l2p.py:59]: `key_similarity` has a `-1 * ` on line 57. Is this your question?

File: src/renate/updaters/experimental/l2p.py [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+        assert isinstance(
+            kwargs["model"], LearningToPromptTransformer
+        ), f"{self.__class__.__name__}  can only train a LearningToPromptTransformer model"
+        f"but got {type(kwargs['model'])}"
+
+        super().__init__(**kwargs)
+        self.prompt_sim_loss_weight = prompt_sim_loss_weight
+        self._loss_collections["train_losses"].update({"key_sim_loss": torchmetrics.MeanMetric()})
+
+    def training_step(
```
[the remark this answers is not in the document]
Comment (prabhuteja12) [reply src/renate/updaters/experimental/l2p.py:88]: The comment on line 91 details why I do this reimplementation. 

TLDR: i need two forward calls. I can possibly hack together something by calling `super().forward` for each of them, but the hassle of managing the logging (because Offline-ER has its own) makes it not worth it.

File: src/renate/benchmark/experiment_config.py [hunk 14]
```diff
@@ -335,7 +344,7 @@ def train_transform(dataset_name: str, model_name: Optional[str] = None) -> Opti
     if dataset_name == "yearbook":
         if (
             model_name is not None
-            and model_name.startswith("VisionTransformer")
+            and "VisionTransformer" in model_name
```
[the remark this answers is not in the document]
Comment (prabhuteja12) [reply src/renate/benchmark/experiment_config.py:347]: Fixed in 03d046b.
